usrsctplib/netinet/sctp_callout.c

```diff
@@ -75,72 +76,201 @@ uint32_t sctp_get_tick_count(void) {
+	}
+#endif
+}
+#endif
+
+static int
+sctp_os_timer_cancel_impl(sctp_os_timer_t* c) {
+	// Assume SCTP_TIMERQ_LOCK done by caller
+
+	c->c_flags &= ~(SCTP_CALLOUT_ACTIVE);
+	c->c_flags |= (SCTP_CALLOUT_CANCELLED);
+
+	if (c->c_flags & SCTP_CALLOUT_PENDING)
+	{
+		c->c_flags &= ~SCTP_CALLOUT_PENDING;
+		sctp_binary_heap_remove(&SCTP_BASE_INFO(timers_queue), &c->heap_node);
+		SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": cancelled pending callout %p\n",
+			__func__, ticks, c);
+		return (1);
+	}
+	SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": cancelled non-pending callout %p\n",
+		__func__, ticks, c);
+	return (0);
+}
+
+void
+sctp_os_timer_describe(const sctp_os_timer_t* t,
+                       size_t max_len,
+                       char* buffer) {
+	snprintf(buffer, max_len, "t=%" PRIu32 ",a=%p", t->c_time, (void*)t);
+}
+
+int
+sctp_os_timer_compare(const sctp_os_timer_t* a, const sctp_os_timer_t* b) {
+	if (a->c_time == b->c_time) {
+		return 0;
+	}
+	return SCTP_UINT32_GT(a->c_time, b->c_time) ? 1 : -1;
+}
 
 void
 sctp_os_timer_init(sctp_os_timer_t *c)
 {
+	SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": request to init callout %p\n", 
+		__func__, sctp_get_tick_count(), c);
 	memset(c, 0, sizeof(*c));
+	sctp_binary_heap_node_init(&c->heap_node, c);
+}
+
+int
+sctp_os_timer_is_pending(const sctp_os_timer_t* c) {
+	SCTP_TIMERQ_LOCK();
+	const int is_pending = (c->c_flags & SCTP_CALLOUT_PENDING);
+	SCTP_TIMERQ_UNLOCK();
+	return is_pending;
+}
+
+int
+sctp_os_timer_is_active(const sctp_os_timer_t* c) {
+	SCTP_TIMERQ_LOCK();
+	const int is_active = (c->c_flags & SCTP_CALLOUT_ACTIVE);
+	SCTP_TIMERQ_UNLOCK();
+	return is_active;
+}
+
+void
+sctp_os_timer_deactivate(sctp_os_timer_t* c) {
+	SCTP_TIMERQ_LOCK();
+	SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": request to deactivate callout %p\n", 
+		__func__, ticks, c);
+	c->c_flags &= ~SCTP_CALLOUT_ACTIVE;
+	SCTP_TIMERQ_UNLOCK();
 }
 
 void
 sctp_os_timer_start(sctp_os_timer_t *c, uint32_t to_ticks, void (*ftn) (void *),
                     void *arg)
 {
 	/* paranoia */
-	if ((c == NULL) || (ftn == NULL))
-	    return;
+	if ((c == NULL) || (ftn == NULL)) {
+		KASSERT(0, ("Timer is NULL or callback is NULL"));
+		return;
+	}
 
 	SCTP_TIMERQ_LOCK();
+	SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": request to start callout %p with delay of %" PRIu32 " ticks\n", 
+		__func__, ticks, c, to_ticks);
 	/* check to see if we're rescheduling a timer */
 	if (c == sctp_os_timer_current) {
 		/*
 		 * We're being asked to reschedule a callout which is
 		 * currently in progress.
 		 */
-		if (sctp_os_timer_waiting) {
+		if ((c->c_flags & SCTP_CALLOUT_CANCELLED) != 0) {
 			/*
 			 * This callout is already being stopped.
 			 * callout.  Don't reschedule.
 			 */
+			SCTPDBG(SCTP_DEBUG_TIMER2, "%s: now=%" PRIu32 ": callout %p is cancelled\n",
+				__func__, ticks, c);
 			SCTP_TIMERQ_UNLOCK();
 			return;
 		}
+		if ((c->c_flags & SCTP_CALLOUT_ACTIVE) == 0) {
```

Comment:
Note for reviewers:
This is not strictly necessary, but it is added as mean of extra debugging of timers